[[ Hooks ]] Add ability to hook into certain aspects of engine functionality.

runrevmark opened the pull request:
The MCHook API is a simple way for parts of the engine to be extended. Currently extra tables of 'global handlers' and 'native controls' can be added.

As it stands this integration still needs to be linked into the main engine - it isn't a clean enough API for public use.

runrevmark committed:
[[ GYP ]] Use correct plist for iOS standalone app bundle

runrevmark committed:
[[ Hooks ]] Simple hook abstraction with infrastructure to hook 'global handlers' and 'native controls'.

runrevmark committed:
[[ Hooks ]] Move global handlers and native controls into hook system.

runrevmark committed:
[[ Engine ]] Make sure iOS engine returns an error if no stack specified in debug mode.

runrevmark committed:
[[ GYP ]] Update the android platform to 10 (2.3.3+)

runrevmark committed:
[[ Engine ]] Add 'attachment' abstraction to MCStack to allow objects to register for interest in window-related events.

runrevmark committed:
[[ Engine ]] Add more methods for native control action definitions
[[ Mobile ]] Make sure iphone/mobileControlDo actions happen on script thread initially.

runrevmark committed:
[[ Engine ]] Refactor MCNativeControlDo.

runrevmark committed:
[[ Engine ]] Add 'waitable' field to native control actions.

runrevmark committed:
[[ Engine ]] Add onPause/onResume to NativeControl module on Android.

runrevmark committed:
[[ Engine ]] Tweaks to allow distinct hooks to work.

runrevmark committed:
[[ Engine ]] Tweaks to native control hooks.

livecodefraser commented on config.sh:
```diff
@@ -266,7 +266,7 @@ WIN_PERL=${WIN_PERL:-"C:/perl/bin/perl.exe"}
 if test "${OS}" = "android" ; then
     ANDROID_NDK_VERSION=${ANDROID_NDK_VERSION:-r10d}
-    ANDROID_PLATFORM=${ANDROID_PLATFORM:-android-8}
+    ANDROID_PLATFORM=${ANDROID_PLATFORM:-android-10}
```
Although we don't have very many (any?) users of Android 2.2, a release note would be useful to say that we now require 2.3.3.

livecodefraser commented:
@livecode-vulcan review okay e45d77d

Marking as reviewed because it looks not-totally-unreasonable (release note excepted) and it could do with a compile test ;)

livecode-vulcan commented:
💙 review by @livecodefraser ok e45d77dffb2f21e3d639dbb5587f41f1ca79f0cb

peter-b commented on engine/src/exec-nativecontrol.cpp:
```diff
@@ -613,7 +613,7 @@ void MCNativeControlExecCreateControl(MCExecContext& ctxt, MCStringRef p_type_na
     
     MCNativeControlType t_type;
-    if (!MCNativeControl::LookupType(p_type_name, t_type))
+    if (!MCLookupNativeControlType(p_type_name, (intenum_t&)t_type))
```
Maybe these would be better with `reinterpret_cast`?

peter-b commented:
In general, this changeset has a very unfortunate lack of comments that describe what the various newly-introduced types and functions and methods and stuff do and why they do it that way. And cryptic commit messages. :sob:

livecode-vulcan commented:
😎 test success e45d77dffb2f21e3d639dbb5587f41f1ca79f0cb
- try-linux-x86: [success](https://vulcan.livecode.com/builders/try-linux-x86/builds/740)
- try-win-x86: [success](https://vulcan.livecode.com/builders/try-win-x86/builds/800)
- try-ios: [success](https://vulcan.livecode.com/builders/try-ios/builds/690)
- try-android-armv6: [success](https://vulcan.livecode.com/builders/try-android-armv6/builds/754)
- try-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-linux-x86_64/builds/685)
- try-mac: [success](https://vulcan.livecode.com/builders/try-mac/builds/677)

runrevmark committed:
[[ Git ]] Make sure .inc files use LF endings.

runrevmark committed:
Merge remote-tracking branch 'origin/develop-7.0' into feature-hooks

runrevmark committed:
[[ GYP ]] Java source files don't appear in projects.

The list of Java source files used on Android has been added to the kernel source file list.
This has no effect on any platform, but is useful for displaying in Xcode and VS.

runrevmark committed:
[[ GYP ]] Remove libs from Win32 kernel link settings.

Libraries which are not needed in the community build have been removed
from the dependent libs section of the Win32 kernel target.

runrevmark committed:
[[ Hooks ]] Generalize Android native control creation code.

The methods such as 'createBrowserControl' in Engine have been refactored to use a single
'createNativeControl' method in the NativeControlModule. This method takes the fully-
qualified name of the class that should be instantiated to create the native control. The
class must derive from NativeControl and have a constructor taking a NativeControlModule
as an argument.

runrevmark committed:
[[ Hooks ]] Correct constructor signature searched for when creating a NativeControl.

runrevmark committed:
[[ Hooks ]] Correct mistyped variable (r_control should have been t_control).

runrevmark committed:
[[ Hooks ]] Add 'signature' field to MCNativeControlActionInfo and use that to dispatch.

The MCNativeControlActionDo method relies on finding an appropriate MCNativeControlActionInfo
struct. The latter describe specific actions by mapping a tag (MCNativeControlAction) to a
method pointer.

The code has been updated to add an explicit signature field to the Info struct which is
use in MCNativeControlDo to dispatch properly. As most actions share similar signatures
this makes much more sense.

runrevmark committed:
[[ Hooks ]] Fix AndroidBrowser exec mappings.

runrevmark committed:
[[ Hooks ]] Factor out core platform message dispatch function.

livecodeian committed:
[[ Camera ]] Add utility foundation function to clear structure memory

runrevmark commented:
@livecode-vulcan review ok 2116c42

livecode-vulcan commented:
💙 review by @runrevmark ok 2116c4249ed71e8514d288c508d83c8788679da1

livecode-vulcan commented:
😎 test success 2116c4249ed71e8514d288c508d83c8788679da1
- try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/7)
- try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/7)
- try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/7)
- try-community-win-x86: [success](https://vulcan.livecode.com/builders/try-community-win-x86/builds/7)
- try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/7)
- try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/7)
